Foundation/NSOrderedSet.swift

```diff
@@ -492,39 +493,33 @@ extension NSMutableOrderedSet {
     }
 
     public func removeAllObjects() {
-        _storage.removeAll()
-        _orderedStorage.removeAll()
+        _storage.removeAllObjects()
+        _orderedStorage.removeAllObjects()
     }
     
     open func remove(_ val: Any) {
-        let object = _SwiftValue.store(val)
-        
-        _storage.remove(object)
-        _orderedStorage.remove(at: index(of: val))
+        _storage.remove(val)
+        _orderedStorage.removeObject(at: index(of: val))
     }
 
     open func removeObjects(in array: [Any]) {
         array.forEach(remove)
     }
     
     open func intersect(_ other: NSOrderedSet) {
-        for item in self where !other.contains(item) {
-            remove(item)
-        }
+        intersectSet(other.set)
     }
 
     open func minus(_ other: NSOrderedSet) {
-        for item in other where contains(item) {
-            remove(item)
-        }
+        minusSet(other.set)
     }
 
     open func union(_ other: NSOrderedSet) {
         other.forEach(add)
     }
     
     open func intersectSet(_ other: Set<AnyHashable>) {
-        for case let item as AnyHashable in self where !other.contains(item) {
+        for case let item as AnyHashable in self.reversed() where !other.contains(item) {
```

Comment:
i assume you're doing this for performance reasons?    there are still O(k) 'shifts'  of the array to fill the holes from the removals.  if you're going for clarity i'd remove the .reversed()  if you're going for clarity combine the shift with a single O(n) pass.   Something like:

```
// pseudo code
i = 0
while i < count{
    nextIndex = i + 1
   if !other.contains(_ordereStorage[i]) {
    if (i+1 < count) {
    _orderedStorage[i] = _orderedStorage[i + 1]
   } else {
      _orderedStorage.removeLastObject()
   }
   // remove from the set too
  } else {
    i = i + 1 // move onto the next index
  }
}

```

Reply:
Doing it without `reversed()` resulted in an index out of bounds error, my guess was because the enumerator was not keeping track of mutations to the array. I could definitely take some time to rewrite this to improve performance

Reply:
Here's a similar method in `NSMutableArray` for reference:

```swift
open func removeObjects(in otherArray: [Any]) {
    let set = Set(otherArray.map { $0 as! AnyHashable } )
    for idx in (0..<count).reversed() {
        if let value = object(at: idx) as? AnyHashable {
            if set.contains(value) {
                removeObject(at: idx)
            }
        }
    }
}
```

Maybe that could be improved a bit as well

Reply:
I think I agree that array could be improved as well.  :) something for the future

Reply:
I have updated this to use an implementation that is based on your pseudocode - I think it could definitely be adapted to NSArray too!